### PR TITLE
feat: Adding Critic module and associated functions 

### PR DESCRIPTION
**Pull request by MarIniOnz**
Adding the Critic Module which discriminates between real and synthetic data.

### CODE REVIEW

File: medmodels/data_synthesis/mtgan/model/critic/critic.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+        The Critic tries to distinguish between real data and synthetic data generated
+        by the Generator. The Critic is trained to maximize the Wasserstein distance
+        between the real and synthetic data distributions.
+
+        Args:
+            real_gru (RealGRU): Real GRU model
+            total_number_of_concepts (int): Size of input features (number of codes)
+            hyperparameters (TrainingHyperparameters): Hyperparameters for training
+            device (torch.device): Device to run the model on
```
Comment (LauraBoenchenLB) [inline medmodels/data_synthesis/mtgan/model/critic/critic.py:46]:
```suggestion
        Args:
            real_gru (RealGRU): Real GRU model.
            total_number_of_concepts (int): Size of input features (number of codes).
            hyperparameters (TrainingHyperparameters): Hyperparameters for training.
            device (torch.device): Device to run the model on.
```

Comment (LauraBoenchenLB) [reply medmodels/data_synthesis/mtgan/model/critic/critic.py:46]: All the docstrings

File: medmodels/data_synthesis/mtgan/model/critic/loss.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            synthetic_hiddens (torch.Tensor): Synthetic hidden states from SyntheticGRU
+                with the same shape as real hiddens
+            number_windows_per_patient (torch.Tensor): number of windows per patient
+
+        Returns:
+            torch.Tensor: gradient penalty
+        """
+        batch_size = len(real_data)
+
+        # Ensure no gradients are calculated for this section to save memory and compute
```
Comment (LauraBoenchenLB) [inline medmodels/data_synthesis/mtgan/model/critic/loss.py:93]:
```suggestion
        # Ensure no gradients are calculated for this section to save memory and computational power.
```

?

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Few comments, mostly just about doc strings. Critic is looking great!!

File: medmodels/data_synthesis/mtgan/model/critic/critic.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            loss_iteration, wasserstein_distance_iteration = (
+                self._train_critic_iteration(
+                    real_data, number_of_windows_per_patient, generator, target_concepts
+                )
+            )
+            loss += loss_iteration
+            wasserstein_distance += wasserstein_distance_iteration
+        loss /= self.critic_iterations
+        wasserstein_distance /= self.critic_iterations
+        self.scheduler.step()
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/critic/critic.py:207]: a bit more spacing

File: medmodels/data_synthesis/mtgan/model/critic/critic.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            loss = 0
+            for data, number_windows_per_patient in test_data:
+                hidden_states = self.real_gru.calculate_hidden_states(
+                    data, number_windows_per_patient
+                )
+                loss += (
+                    self(data, hidden_states, number_windows_per_patient).mean().item()
+                )
+            loss = -loss / len(test_data)
+            return loss
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/critic/critic.py:232]: a bit more spacing

File: medmodels/data_synthesis/mtgan/model/critic/loss.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    def __init__(self, critic: Critic, lambda_: float) -> None:
+        """Constructor for the Critic Wasserstein-Loss with gradient penalty.
+
+        Args:
+            critic (Critic): Critic model.
+            lambda_ (int): Gradient penalty coefficient.
+        """
+        super().__init__()
+        self.critic = critic
+        self.lambda_ = lambda_
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/critic/loss.py:26]: could there be a better name to use here instead of `lambda` so you don't have to add the `_`? If nothing comes to mind then that's fine :)

File: medmodels/data_synthesis/mtgan/model/critic/loss.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        gradient_penalty = self.compute_gradient_penalty(
+            real_data,
+            real_hiddens,
+            synthetic_data,
+            synthetic_hiddens,
+            number_windows_per_patient,
+        )
+        wasserstein_distance = critic_real.mean() - critic_synthetic.mean()
+        critic_loss = -wasserstein_distance + gradient_penalty
+        return critic_loss, wasserstein_distance
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/critic/loss.py:65]: a bit more spacing

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
some teeny tiny changes requested

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
one smol thing

**Review by JabobKrauskopf — APPROVED** [review]
image

**Review by LauraBoenchenLB — APPROVED** [review]
Let's gooooo MTGAN 🦭
